src/main/java/core/AbstractGameState.java

```diff
@@ -53,7 +54,7 @@ public abstract class AbstractGameState {
     protected ElapsedCpuChessTimer[] playerTimer;
 
     // A record of all actions taken to reach this game state
-    private List<AbstractAction> history = new ArrayList<>();
+    private List<Pair<Integer, AbstractAction>> history = new ArrayList<>();
```

Comment:
Just one small comment, if we can update the comment for what this history is to indicate what's recorded for the pair (not immediately obvious).